Adding fuzzy search functionality to front end search

tcollins2011 opened the pull request:
Why: hackathon project: When searching for a tool using the front end search, the user should be able to make small typos and still receive search results.

What: On tool search an exact substring match is checked to see if a tool matches. If this fails, a modified form of the Damerau–Levenshtein distance algorithm is applied to to find similar results. 

[Galaxy.webm](https://user-images.githubusercontent.com/62220253/228048996-24b3889a-1bbe-4c08-8d37-8d55926edcef.webm)


## How to test the changes?

- [ ] Instructions for manual testing are as follows:
  1. Click the tool search bar to enter a search
  2. Search for a tool name with both correct and incorrect spelling to see how it appears

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

tcollins2011 committed:
Added in fuzzy search to light weight front end tool search

assuntad23 reviewed (commented):
Nice work!

dannon commented on client/src/components/Panels/utilities.js:
```diff
@@ -98,18 +98,56 @@ export function searchToolsByKeys(tools, keys, query) {
     return orderBy(returnedTools, ["order"], ["desc"]).map((tool) => tool.id);
 }
 
+export function dLDistance(query, value) {
```
Instead of reimplementing this here, I'd suggest consolidation with https://github.com/galaxyproject/galaxy/blob/dev/client/src/utils/levenshtein.js

edit: it may well be worth updating *that*, as the code is fairly old there, but we should have a single levenshteinDistance function that we reuse

tcollins2011 replied:
I'll look in to updating the levensnshtein.js function and will possibly add the new Damerau–Levenshtein distance function if it isn't there. Also if someone has the time I changed the functionality of the search again to better deal with substring partial matches. Ie) if someone misspells "filtar" filter will still show up and all results that also contain the substring filter.

tcollins2011 committed:
added in substring partial match functionality to fuzzy search

tcollins2011 committed:
chaning threshold variables

tcollins2011 committed:
cleaned up substring search and tightened threshold values

tcollins2011 committed:
updated levenshtein.js to include transpositions and restructured utilites to use the new location for the levensthein distance algorithm

tcollins2011 committed:
starting to add error detection

tcollins2011 committed:
removed did you mean prompt from this branch

tcollins2011 committed:
changed code to fix linter conflicts and applied prettier styling

ahmedhamidawan committed:
fuzzy search - Only include non DL results, add client tests

tcollins2011 committed:
refactored dLDistance and fixed insertion bug

tcollins2011 committed:
recommented refactored code

tcollins2011 committed:
fixed deletion error

ahmedhamidawan committed:
fuzzy search - only ever run `dLDistance()` if no searches are found with `.match()`

ahmedhamidawan committed:
fuzzy search - reorganize code

ahmedhamidawan committed:
fuzzy search - run DL separately for name and other keys

ahmedhamidawan committed:
fuzzy search - use vanilla JS Worker to isolate ToolSearch in a thread

ahmedhamidawan committed:
fuzzy search - Add `Did you mean` feature

guerler commented on client/src/components/Panels/Common/ToolSearch.vue:
```diff
@@ -82,8 +87,12 @@ export default {
     },
+    beforeDestroy() {
+        this.searchWorker.terminate();
+    },
```
Notice the linting issue below. Moving the compute property up will resolve it.

guerler commented on client/src/components/Panels/Common/ToolSearch.vue:
```diff
@@ -101,20 +110,41 @@ export default {
     },
+    mounted() {
+        this.searchWorker = new Worker(new URL("../toolSearch.worker.js", import.meta.url));
+        this.searchWorker.onmessage = (event) => {
```
Minor but here you can write `{ data }` instead of `event`.

guerler commented on client/src/components/Panels/Common/ToolSearch.vue:
```diff
@@ -101,20 +110,41 @@ export default {
         },
     },
+    mounted() {
+        this.searchWorker = new Worker(new URL("../toolSearch.worker.js", import.meta.url));
```
Moving this into a separate module which creates and configures the worker might simplify the usability here. Thanks a lot for working on this. This overall looks really good to me.

ahmedhamidawan committed:
fix ordering of properties in ToolSearch.vue

ahmedhamidawan committed:
sanitize ToolSearch query, fixed-width icons

ahmedhamidawan committed:
move tool fuzzy filtering into a composable

remove search test from ToolBox (worker not producing results in jest)

remove composable, use worker in vue comp

ahmedhamidawan committed:
fix jest fail for added owner field
